### PR TITLE
Replace AsFixedBytes with Sized + AsRef<[u8]>

### PR DESCRIPTION
**Pull request by Latrasis**
Implementation for #34

### CODE REVIEW

File: core/src/ser.rs [hunk 1]
```diff
@@ -140,14 +132,14 @@ pub trait Writer {
 
 	/// Writes a variable number of bytes. The length is encoded as a 64-bit
 	/// prefix.
-	fn write_bytes(&mut self, bytes: &AsFixedBytes) -> Result<(), Error> {
-		try!(self.write_u64(bytes.as_fixed_bytes().len() as u64));
+	fn write_bytes<T: Sized + AsRef<[u8]>>(&mut self, bytes: &T) -> Result<(), Error> {
```
Comment (Latrasis) [inline core/src/ser.rs]: Note that this makes the `Writer` trait generic.

**Comment by ignopeverell** [comment]
Instead of defining <T: Sized + AsRef<[u8]>> on everything (especially outside of the module, like in ser.rs), I'd still introduce a separate type, maybe something like (not set on the name, if you can find something better):

    trait SizedBytes: Sized + AsRef<[u8]>

That should help encapsulating it a bit.

Also in secp256k1zkp I think it makes sense to implement AsRef as part of the impl_array_newtype macro.

**Comment by Latrasis** [comment]
Considering that `[u8]` is unsized, why is it being included as part of `AsFixedBytes`? I'm having problems currently with using implementing it as a marker trait, because `[u8]` is unsized.

File: core/src/core/hash.rs [hunk 2]
```diff
@@ -121,7 +121,7 @@ impl Hashed for [u8] {
 	fn hash(&self) -> Hash {
 		let mut hasher = HashWriter::default();
 		let mut ret = [0; 32];
-		ser::Writer::write_fixed_bytes(&mut hasher, &self).unwrap();
+		ser::Writer::write_fixed_bytes(&mut hasher, &self.to_owned()).unwrap();
```
Comment (Latrasis) [inline core/src/core/hash.rs]: Had to box the slice to have it `Sized`. This doesn't seem like a solution though...

**Comment by ignopeverell** [comment]
Can you describe to problem maybe, see if I can help (or if my review comments were just wrong)? :)

**Comment by Latrasis** [comment]
`impl Hashed for [u8]` relies on `ser::Writer::write_fixed_bytes` however `self` (which is `[u8]` in this case) can't be`AsFixedBytes`, because any `[T]` is not `Sized`... So i'm currently slightly stuck on this because i'm not sure what solution would be a better fit. The previous `FixedBytes` seems to have been contradictory on `impl FixedBytes for [u8]` because slices can't be guaranteed to have fixed size. Perhaps you mean something else than `Sized`?

**Comment by ignopeverell** [comment]
I pushed a change that removes `impl Hashed for [u8]`, it was really only there to be able to hash an empty array. In your branch you should be able to remove the `AsFixedBytes` implementations for `[u8]` and `String` as they're only required for `Hash`, which in your branch should implement `AsRef`. I believe that should help.

**Comment by Latrasis** [comment]
Thank you!

**Comment by ignopeverell** [comment]
Are you ok with where the pull request is at now? If so I'll go ahead and merge it, it looks fine to me.

**Comment by Latrasis** [comment]
Yup

**Pull request merged**